app/gui2/src/components/visualizations/JSONVisualization.vue

```diff
@@ -5,14 +5,226 @@ export const inputType = 'Any'
+  childCount(): number {
+    let count = 0
+    for (const _child of this.children()) count += 1
+    return count
+  }
+}
+
+class JsonObject extends JsonNode {
+  public declare readonly data: Record<string, Json>
+
+  constructor(
+    data: Record<string, Json>,
+    parent: { node: JsonNode; selector: string | number } | undefined,
+  ) {
+    super(data, parent)
+  }
+
+  *children(): IterableIterator<JsonNode & ChildOfObject> {
+    for (const [key, value] of Object.entries(this.data))
+      yield JsonNode.new(value, { node: this, selector: key }) as JsonNode & ChildOfObject
+  }
+
+  keys(): string[] {
+    return Object.keys(this.data)
+  }
+}
+interface ChildOfObject {
+  readonly parent: { node: JsonNode; selector: string }
+}
+
+class JsonArray extends JsonNode {
+  public declare readonly data: Json[]
+
+  constructor(data: Json[], parent: { node: JsonNode; selector: string | number } | undefined) {
+    super(data, parent)
+  }
+
+  *children(): IterableIterator<JsonNode & ChildOfArray> {
+    let index = 0
+    for (const value of this.data) {
+      yield JsonNode.new(value, { node: this, selector: index }) as JsonNode & ChildOfArray
+      index += 1
+    }
+  }
+
+  childCount(): number {
+    return this.data.length
+  }
+}
+interface ChildOfArray {
+  readonly parent: { node: JsonNode; selector: number }
+}
+
+class JsonPrimitive extends JsonNode {
+  constructor(data: Json, parent: { node: JsonNode; selector: string | number } | undefined) {
+    super(data, parent)
+  }
+}
+
+function render(value: JsonNode, depth: number = 0, parentInline?: boolean) {
+  const html = document.createElement('span')
+  const inline = parentInline || JSON.stringify(value.data).length <= 40
+  if (value instanceof JsonArray) {
+    html.append('[')
+    let i = 0
+    for (const element of value.children()) {
+      if (i !== 0) html.append(',')
+      html.append(inline ? ' ' : newline(depth + 1))
+      const elementHtml = document.createElement('span')
+      elementHtml.classList.add('element')
+      elementHtml.append(render(element, depth + 1, inline))
+      addClickHandler(elementHtml, element)
+      html.append(elementHtml)
+      i += 1
+    }
+    html.append(inline ? ' ' : newline(depth))
+    html.append(']')
+  } else if (value instanceof JsonObject) {
+    html.append('{')
+    let i = 0
+    for (const field of value.children()) {
+      if (i !== 0) html.append(',')
+      html.append(inline ? ' ' : newline(depth + 1))
+      const fieldHtml = document.createElement('span')
+      fieldHtml.classList.add('field')
+      fieldHtml.append(renderKey(field.parent.selector), ': ', render(field, depth + 1, inline))
+      addClickHandler(fieldHtml, field)
+      html.append(fieldHtml)
+      i += 1
+    }
+    html.append(inline ? ' ' : newline(depth))
+    html.append('}')
+  } else {
+    html.classList.add('primitive')
+    html.append(JSON.stringify(value.data))
+  }
+  return html
+}
```

Comment:
Why manual DOM manipulation? This could be a vue template.

Reply:
It definitely should be. Fix incoming...